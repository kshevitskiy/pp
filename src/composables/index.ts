export * from './storage'
export * from './user-list'
