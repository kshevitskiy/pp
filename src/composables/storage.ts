export enum StorageKeys {
  REMOVED_USERS = 'pp:removed-users'
}

export function useStorage() {
  function getItem(key: StorageKeys): string | null {
    return localStorage.getItem(key)
  }

  function setItem(key: StorageKeys, value: any): void {
    localStorage.setItem(key, value)
  }

  function removeItem(key: StorageKeys): void {
    localStorage.removeItem(key)
  }

  return {
    getItem,
    setItem,
    removeItem
  }
}
