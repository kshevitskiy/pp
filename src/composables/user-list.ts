import { useService } from '~/service'
import { computed, ref, watch } from 'vue'
import { StorageKeys, useStorage } from '~/composables'
import { User } from '~/types'
import { delay } from '~/assets/utils'

interface UserListModule {
  list(): Promise<void>
  remove(user: User): void
  reset(): void
}

const { getItem, setItem } = useStorage()

export const users = ref<User[]>([])
export const removedUsers = ref<string[]>([])
export const pending = ref(false)

export const usersCount = computed(() =>
  Array.isArray(users.value) ? users.value.length : 0
)

export function useUserList(): UserListModule {
  const since = 0
  const perPage = 10

  watch(removedUsers.value, update, { deep: true })

  function update(): void {
    users.value = users.value.filter(
      (user) => !removedUsers.value.includes(user.login)
    )
  }

  function setRemovedUsers(value: string[]): void {
    removedUsers.value = value
    setItem(StorageKeys.REMOVED_USERS, JSON.stringify(removedUsers.value))
  }

  async function list(): Promise<void> {
    pending.value = true

    try {
      const response = await useService().users.list({
        per_page: perPage,
        since
      })

      // TODO: Refactor. Delay was added for demonstration purpose and should be removed later
      await delay(1000)
      if (Array.isArray(response.data)) {
        users.value = response.data
      }
    } catch (exception) {
      console.error(exception)
    } finally {
      pending.value = false
    }
  }

  function remove(user: User): void {
    const isRemoved = removedUsers.value.includes(user.login)

    if (!isRemoved) {
      removedUsers.value.push(user.login)
      setRemovedUsers(removedUsers.value)
    }
  }

  async function reset() {
    if (removedUsers.value.length > 0) {
      setRemovedUsers([])
      await list()
    }
  }

  return {
    list,
    remove,
    reset
  }
}

export async function initUserList(): Promise<void> {
  try {
    await useUserList().list()

    const removedUsersCollection = JSON.parse(
      getItem(StorageKeys.REMOVED_USERS) || '[]'
    )

    if (Array.isArray(removedUsersCollection)) {
      removedUsers.value.push(...removedUsersCollection)
    }
  } catch (exception) {
    console.error(exception)
  }
}
