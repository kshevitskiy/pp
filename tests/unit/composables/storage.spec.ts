import { assert, describe, expect, it } from 'vitest'
import { useStorage } from '~/composables'

describe('storage composable', () => {
  it('to have defined methods properties', () => {
    const storage = useStorage()

    expect(storage).toHaveProperty('getItem')
    expect(storage).toHaveProperty('setItem')
    expect(storage).toHaveProperty('removeItem')
  })
})

// TODO: Test each separate composable method
// ...
