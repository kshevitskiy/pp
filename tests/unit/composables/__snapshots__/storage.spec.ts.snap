// Vitest Snapshot v1

exports[`storage composable > snapshot 1`] = `
{
  "foo": "bar",
}
`;

exports[`suite name > snapshot 1`] = `
{
  "foo": "bar",
}
`;
